feat: depth estimation dataset guide. (#5379)
huggingface/datasets: docs/source/depth_estimation.mdx

# Depth estimation

Depth estimation datasets are used to train a model to approximate the relative distance of every pixel in an
image from the camera aka depth. The applications enabled by these datasets mostly lie in areas like visual machine
perception and perception in robotics. This guide will show you how to apply transformations
to a depth estimation dataset.

Before you start, make sure you have up-to-date versions of `albumentations` installed:

```bash
pip install -U albumentations 
```

[Albumentations](https://albumentations.ai/) is a Python library for performing data augmentation
for computer vision. It supports various computer vision tasks such as image classification, object
detection, segmentation, and keypoint estimation.

This guide uses the [NYU Depth V2](https://huggingface.co/datasets/sayakpaul/nyu_depth_v2) dataset which is 

> comprised of video sequences from a variety of indoor scenes as recorded by both the RGB and Depth cameras
> from the Microsoft [Kinect](http://www.xbox.com/kinect). The dataset consists of scenes from 3 cities and provides images along with

Reviewer: I don't think it's necessary or relevant to mention here that it was recorded from a Kinect. A link to the dataset card is fine :)

```suggestion
comprised of video sequences from various indoor scenes, recorded by RGB and depth cameras. The dataset consists of scenes from 3 cities and provides images along with
```

Author: Agreed. Thanks.

their depth maps as labels.

Load the `train` split of the dataset and take a look at an example:

```py
>>> from datasets import load_dataset

>>> train_dataset = load_dataset("sayakpaul/nyu_depth_v2", split="train")
>>> index = 17
>>> dataset[index]
{'image': <PIL.PngImagePlugin.PngImageFile image mode=RGB size=640x480>,
 'depth_map': <PIL.PngImagePlugin.PngImageFile image mode=L size=640x480>}
```

> The dataset has three fields:

Reviewer: ```suggestion
The dataset has two fields:
```

* `image`: a PIL image object.
* `depth_map`: depth map of the image.

Next, check out an image with:

```py
>>> train_dataset[index]["image"]
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_sample.png">
</div>

> Similarly, you can check out the respective depth map:

Reviewer: Probably better if we tell the user to look at the depth map instead of suggesting it

```suggestion
Now take a look at its corresponding depth map:
```

```py
>>> train_dataset[index]["depth_map"]
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_target.png">
</div>

> It's all black! You need to apply some preprocessing to the depth map to properly visualize it. 

Reviewer: I would delete the **note** and combine the code snippets. 

```suggestion
It's all black! You'll need to add some color to the depth map to visualize it properly (the utility below is taken from the [FastDepth repository](https://github.com/dwofk/fast-depth/blob/master/utils.py)).
```

Author: See if it looks better now.

>>> d_min = np.min(depth_target)
>>> d_max = np.max(depth_target)
>>> depth_target = colored_depthmap(depth_target, d_min, d_max)

>>> plt.imshow(depth_target.astype("uint8"))
>>> plt.axis("off")
>>> plt.show()
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_target_viz.png">
</div>

Now apply some augmentations with `albumentations`. The augmentation transformations include:

* Random horizontal flipping
* Random cropping 
* Random brightness and contrast 
* Random gamma correction 
* Random hue saturation

```py 
>>> import albumentations as A

>>> crop_size = (448, 576)
>>> transforms = [
...     A.HorizontalFlip(),
...     A.RandomCrop(crop_size[0], crop_size[1]),
...     A.RandomBrightnessContrast(),
...     A.RandomGamma(),
...     A.HueSaturationValue()
... ]
```

Since these transformations (such as flipping, cropping) also impact the depth maps, you need to define a mapping
while applying these transformations:

```py 
>>> additional_targets = {"depth": "mask"}
>>> aug = A.Compose(transforms=transforms, additional_targets=additional_targets)
```

> This will ensure that you can pass an image and its corresponding depth map to `aug`.

Reviewer: ```suggestion
This ensures you can pass an image and its corresponding depth map to `aug`.
```

Create a function to apply the transformation to the images as well as their depth maps:

```py 
>>> def transforms(examples):
...     transformed_images, transformed_maps = [], []
...     for image, depth_map in zip(examples["image"], examples["depth_map"]):
...         image, depth_map = np.array(image), np.array(depth_map)
...         transformed = aug(image=image, depth=depth_map)
...         transformed_images.append(transformed["image"])
...         transformed_maps.append(transformed["depth"])
...
...     examples["pixel_values"] = transformed_images
...     examples["labels"] = transformed_maps
...     return examples
```

Use the [`~Dataset.set_transform`] function to apply the transformation on-the-fly to batches of the dataset to consume less disk space:

```py
>>> train_dataset.set_transform(transforms)
```

> You can verify the transformation worked by indexing into the `pixel_values` and `labels` of an example:

Reviewer: I would combine the two code snippets below.

```suggestion
You can verify the transformation worked by indexing into the `pixel_values` and `labels` of an example image:
```

Author: The succession made it visually simpler to read through. Hence decided to keep it that way.

```py
>>> image = np.array(train_dataset[index]["pixel_values"])
>>> depth_map = np.array(train_dataset[index]["labels"])
```

```py
>>> plt.imshow(image)
>>> plt.axis("off")
>>> plt.show()
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_sample_aug.png">
</div>

> ```py 

Reviewer: Maybe add some text here that says something like:

`Visualize the same transformation on the image's corresponding depth map:`

```py 
>>> import numpy as np
>>> import matplotlib.pyplot as plt

>>> cmap = plt.cm.viridis

>>> def colored_depthmap(depth, d_min=None, d_max=None):
...     if d_min is None:
...         d_min = np.min(depth)
...     if d_max is None:
...         d_max = np.max(depth)
...     depth_relative = (depth - d_min) / (d_max - d_min)
...     return 255 * cmap(depth_relative)[:,:,:3]

>>> def show_depthmap(depth_map):
...    if not isinstance(depth_map, np.ndarray):
...        depth_map = np.array(depth_map)
...    if depth_map.ndim == 3:
...        depth_map = depth_map.squeeze()

...    d_min = np.min(depth_map)
...    d_max = np.max(depth_map)
...    depth_map = colored_depthmap(depth_map, d_min, d_max)

...    plt.imshow(depth_map.astype("uint8"))
...    plt.axis("off")
...    plt.show()

>>> show_depthmap(train_dataset[index]["depth_map"])
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_target_viz.png">
</div>

> You can also visualize a collage of different images and the corresponding depth maps.

Reviewer: I think the wording might be a little clearer this way. Typically `collage` has a very specific reference to art. 

```suggestion
You can also visualize several different images and their corresponding depth maps.
```

```py
>>> def merge_into_row(input_image, depth_target):
...     if not isinstance(input_image, np.ndarray):
...         input_image = np.array(input_image)
...
...     d_min = np.min(depth_target)
...     d_max = np.max(depth_target)
...     depth_target_col = colored_depthmap(depth_target, d_min, d_max)
...     img_merge = np.hstack([input_image, depth_target_col])
...
...     return img_merge

>>> random_indices = np.random.choice(len(train_dataset), 9).tolist()
>>> plt.figure(figsize=(15, 6))
>>> for i, idx in enumerate(random_indices):
...     ax = plt.subplot(3, 3, i + 1)
...     image_viz = merge_into_row(
...         train_dataset[idx]["image"], train_dataset[idx]["depth_map"]
...     )
...     plt.imshow(image_viz.astype("uint8"))
...     plt.axis("off")
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/datasets/depth_est_collage.png">
</div>

Now apply some augmentations with `albumentations`. The augmentation transformations include:

* Random cropping 
* Random brightness and contrast 
* Random gamma correction 

> * Random hue saturation

Author: I had to drop horizontal flipping. Here's why. 

The behavior seems to be unstable with `datasets` when applied via `set_transform()`. Consider the following as an example. 

<img width="1533" alt="debug_augmentations" src="https://user-images.githubusercontent.com/22957388/209280992-08154743-9855-4eb2-842a-dab25432072a.png">

You'll notice that subplots tagged as "wrong" have flipped images but their corresponding depth maps are not flipped. This is not a consistent case, though. Subplots tagged as "right", for example. You'll notice that the images and their depth maps have both been flipped there (compare them to the originals and you should be able to spot the difference). 

I have verified if there's a problem with `albumentations` but that doesn't seem to be the case. Refer to this notebook: https://gist.github.com/sayakpaul/63ae5f7971f634894ab37e76e6b3879c. You'll notice that I took individual samples from the dataset and passed them through the `aug` chain and the results were fine. 

Have you experienced this before? Is anything obviously wrong you can spot in the code? 

Let me know if anything's unclear. 

@nateraw @lhoestq

Reviewer 2: I'm unable to reproduce this issue 🙁 . Take a look at this [simplified notebook](https://gist.github.com/nateraw/434fd94de897d76f449a805fcc3ec623).

Author: Yes @nateraw. But it doesn't seem to be consistent, as I stated here:

> You'll notice that subplots tagged as "wrong" have flipped images but their corresponding depth maps are not flipped. This is not a consistent case, though. Subplots tagged as "right", for example.

Sometimes, the augmented depth maps are correct, i.e., the flip transformation has been correctly applied, and sometimes, that doesn't seem to be the case. You'll notice in the collage [here](https://github.com/huggingface/datasets/pull/5379#discussion_r1056065300).

Reviewer 2: Yea I noticed that, but I ran it over and over and never got that behavior. Do you have a minimal reproduction of the behavior?

Author: I don't. Strangely, it only comes up when I try to put it inside a collection of images.

I will try to reproduce the behavior minimally and let you know here.

Author: My god! Thanks for saving me!
